:sparkles: KCP adopts existing machines

sethp-nr opened the pull request:
<!-- please add a icon to the title of this PR (see https://sigs.k8s.io/cluster-api/VERSIONING.md), and delete this line and similar ones -->
<!-- the icon will be either ⚠️ (:warning:, major or breaking changes), ✨ (:sparkles:, minor or feature additions), 🐛 (:bug:, patch and bugfixes), 📖 (:book:, documentation or proposals), or 🏃 (:running:, other) -->

**What this PR does / why we need it**:

A mostly-functional prototype for having the KCP controller identify
Machines that belong to the control plane of an existing cluster and
adopt them.

Additionally, we set set a "best guess" value for the
kubeadm.controlplane.cluster.x-k8s.io/hash on the adopted machine as if
it were generated by a KCP in the past.

Co-authored-by: mnguyen <mnguyen@newrelic.com>


**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #2214


# TODO

The bits of this work that are still in flight are:

- [x] (critical) Handling the adoption of cluster PKI secrets (see #2455)
- [x] (critical) Validating that the machines being adopted are actually Kubeadm-managed
- [x] Re-working the adoption pieces to make better use of the `FilterableMachineCollection`
- [x] Writing safety checks around upgrade versions, i.e.
     - [x] ~a validating webhook that rejects two major version bumps to a KCP~ See #2578
     - [x] a fail-fast path in the adoption if we'd be adopting machines that are "too old")
- [x] Write an e2e test, especially around the Secret adoption behavior (it's kind of a pain to test in unit)

/hold
/milestone v0.3.0

vincepri commented:
/milestone v0.3.0

vincepri commented:
/assign 

to review later

sethp-nr commented:
OK, I think this is RFR – I left the hold label on there so we can make a judgement call about whether it needs to be feature-gated after a review cycle or two.

vincepri commented on controlplane/kubeadm/controllers/kubeadm_control_plane_controller.go:
```diff
@@ -834,3 +845,145 @@ func (r *KubeadmControlPlaneReconciler) ClusterToKubeadmControlPlane(o handler.M
+	// We do an uncached full quorum read against the KCP to avoid re-adopting Machines the garbage collector just intentionally orphaned
+	// See https://github.com/kubernetes/kubernetes/issues/42639
+	uncached := controlplanev1.KubeadmControlPlane{}
+	err := r.uncachedClient.Get(ctx, client.ObjectKey{Namespace: kcp.Namespace, Name: kcp.Name}, &uncached)
```
If we adopt a Machine that's going to be deleted, wouldn't we get a notification and KCP reconciles again afterwards? 

I guess what I'm trying to understand is the actual problem this extra client is trying to solve, the comment explains part of it, but I'm missing why this would be a problem given that the controller will keep reconciling the resource

sethp-nr replied:
There's an in-depth explanation of the flow in the test suite, but the issue has to do with a deleted KCP possibly re-adopting (and then deleting) dependents that we don't want cleaned up.

There is another, different race with orphaning dependents (see #2505) but from my brief survey of k/k's adoption code it all does a check like this one before it actually takes ownership of anything.

detiber replied:
If this is a common pattern for adoption, we likely should create an issue to track standardizing on this approach. I also wonder if it would be worth it to try and get support for this more directly in controller-runtime somehow.

sethp-nr replied:
Good call – post-release when I file the follow-up "let's talk about deletes" issue for 2505 there I'll make sure to include this part.

vincepri commented on controlplane/kubeadm/controllers/kubeadm_control_plane_controller.go:
```diff
@@ -834,3 +845,145 @@ func (r *KubeadmControlPlaneReconciler) ClusterToKubeadmControlPlane(o handler.M
+	uncached := controlplanev1.KubeadmControlPlane{}
+	err := r.uncachedClient.Get(ctx, client.ObjectKey{Namespace: kcp.Namespace, Name: kcp.Name}, &uncached)
+	if err != nil {
+		return fmt.Errorf("can't recheck DeletionTimestamp: %v", err)
```
```suggestion
		return errors.Wrapf(err, "failed to retrieve <...>", ...)
```
To be consistent with other error patterns, let's also add the name and namespace of the object we're trying to get

vincepri commented on controlplane/kubeadm/controllers/kubeadm_control_plane_controller.go:
```diff
@@ -834,3 +845,145 @@ func (r *KubeadmControlPlaneReconciler) ClusterToKubeadmControlPlane(o handler.M
+		return fmt.Errorf("can't recheck DeletionTimestamp: %v", err)
+	}
+	if !uncached.DeletionTimestamp.IsZero() {
+		return fmt.Errorf("%v/%v has just been deleted at %v", kcp.GetNamespace(), kcp.GetName(), kcp.GetDeletionTimestamp())
```
We should use `pkg/errors` everywhere we can

vincepri commented on controlplane/kubeadm/controllers/kubeadm_control_plane_controller.go:
```diff
@@ -834,3 +845,145 @@ func (r *KubeadmControlPlaneReconciler) ClusterToKubeadmControlPlane(o handler.M
+
+		// TODO instead of returning error here, we should instead Event and add a watch on potentially adoptable Machines
+		if ref == nil || ref.Kind != "KubeadmConfig" {
+			return fmt.Errorf("unable to adopt Machine %v/%v: expected a ConfigRef of kind KubeadmConfig but instead found %v", m.Namespace, m.Name, ref)
```
We should use `pkg/errors` everywhere we can

vincepri commented on controlplane/kubeadm/controllers/kubeadm_control_plane_controller.go:
```diff
@@ -834,3 +845,145 @@ func (r *KubeadmControlPlaneReconciler) ClusterToKubeadmControlPlane(o handler.M
+
+		// TODO instead of returning error here, we should instead Event and add a watch on potentially adoptable Machines
+		if ref.Namespace != "" && ref.Namespace != kcp.Namespace {
+			return fmt.Errorf("could not adopt resources from KubeadmConfig %v/%v: cannot adopt across namespaces", ref.Namespace, ref.Name)
```
We should use `pkg/errors` everywhere we can

vincepri commented on controlplane/kubeadm/controllers/kubeadm_control_plane_controller.go:
```diff
@@ -834,3 +845,145 @@ func (r *KubeadmControlPlaneReconciler) ClusterToKubeadmControlPlane(o handler.M
+			return errors.Wrapf(err, "failed to parse kubernetes version %q", *m.Spec.Version)
+		}
+
+		dist := func(a, b uint64) uint64 {
```
What do you think about moving this to the util package and make it work on `semver.Version`?

vincepri commented on controlplane/kubeadm/controllers/kubeadm_control_plane_controller.go:
```diff
@@ -834,3 +845,145 @@ func (r *KubeadmControlPlaneReconciler) ClusterToKubeadmControlPlane(o handler.M
+			return err
+		}
+
+		m.SetOwnerReferences(util.EnsureOwnerRef(m.GetOwnerReferences(), metav1.OwnerReference{
```
Consider switching to controller-runtime's https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/controller/controllerutil/controllerutil.go#L58

vincepri commented on controlplane/kubeadm/controllers/kubeadm_control_plane_controller.go:
```diff
@@ -834,3 +845,145 @@ func (r *KubeadmControlPlaneReconciler) ClusterToKubeadmControlPlane(o handler.M
 	return nil
 }
+
+func (r *KubeadmControlPlaneReconciler) AdoptMachines(ctx context.Context, kcp *controlplanev1.KubeadmControlPlane, machines internal.FilterableMachineCollection) error {
```
```suggestion
func (r *KubeadmControlPlaneReconciler) adoptMachines(ctx context.Context, kcp *controlplanev1.KubeadmControlPlane, machines internal.FilterableMachineCollection) error {
```

vincepri commented on controlplane/kubeadm/controllers/kubeadm_control_plane_controller.go:
```diff
@@ -834,3 +845,145 @@ func (r *KubeadmControlPlaneReconciler) ClusterToKubeadmControlPlane(o handler.M
+	return nil
+}
+
+func (r *KubeadmControlPlaneReconciler) AdoptOwnedSecrets(ctx context.Context, kcp *controlplanev1.KubeadmControlPlane, currentOwner metav1.Object) error {
```
```suggestion
func (r *KubeadmControlPlaneReconciler) adoptOwnedSecrets(ctx context.Context, kcp *controlplanev1.KubeadmControlPlane, currentOwner metav1.Object) error {
```

vincepri commented on controlplane/kubeadm/controllers/kubeadm_control_plane_controller.go:
```diff
@@ -834,3 +845,145 @@ func (r *KubeadmControlPlaneReconciler) ClusterToKubeadmControlPlane(o handler.M
+		}
+
+		ss := corev1.Secret{}
+		s.DeepCopyInto(&ss)
```
```suggestion
		ss := s.DeepCopy()
```

vincepri commented on controlplane/kubeadm/controllers/kubeadm_control_plane_controller.go:
```diff
@@ -834,3 +845,145 @@ func (r *KubeadmControlPlaneReconciler) ClusterToKubeadmControlPlane(o handler.M
+		ss := corev1.Secret{}
+		s.DeepCopyInto(&ss)
+
+		ss.SetOwnerReferences(util.ReconcileOwnerRef(ss.GetOwnerReferences(), metav1.OwnerReference{
```
Consider switching to controller-runtime's https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/controller/controllerutil/controllerutil.go#L58

sethp-nr replied:
That wouldn't work here – we're explicitly taking control away from the current owner, but `SetControllerReference` returns an error if there's already a controller.

vincepri commented on controlplane/kubeadm/controllers/kubeadm_control_plane_controller.go:
```diff
@@ -834,3 +845,145 @@ func (r *KubeadmControlPlaneReconciler) ClusterToKubeadmControlPlane(o handler.M
+		}, currentOwner))
+
+		err := r.Client.Update(ctx, &ss)
+		if err != nil {
```
Any reason not to use Patch instead?

detiber replied:
+1 to using patch here

sethp-nr replied:
The Patch that's implemented is strategic merge. We'd need to use a different strategy to do a targeted remove of the old controller reference before adding the new one.

vincepri commented on controlplane/kubeadm/controllers/kubeadm_control_plane_controller.go:
```diff
@@ -227,13 +231,25 @@ func (r *KubeadmControlPlaneReconciler) reconcile(ctx context.Context, cluster *
 		return ctrl.Result{}, err
 	}
 
+	adoptableMachines := controlPlaneMachines.Filter(internal.AdoptableControlPlaneMachines(cluster.Name))
```
I wonder if this behavior should be opt-in rather than having it run by default, wdyt?

detiber replied:
One would already need to create KCP resource to trigger the adoption, or do you think that an additional safeguard would be needed?

vincepri replied:
Not sure yet, that's a good point though, it's opt-in if you create and link the object

sethp-nr replied:
Yeah, that felt like the opt-in edge to me. Otherwise, the KCP creates a new, separate control plane that's "for" the same cluster as the old one but otherwise independent.

vincepri commented on util/util.go:
```diff
@@ -268,6 +268,22 @@ func EnsureOwnerRef(ownerReferences []metav1.OwnerReference, ref metav1.OwnerRef
 	return ownerReferences
 }
 
+// ReconcileOwnerRef re-parents an object from one OnwerReference to another
```
```suggestion
// ReconcileOwnerRef re-parents an object from one OwnerReference to another
```

vincepri commented on util/util.go:
```diff
@@ -268,6 +268,22 @@ func EnsureOwnerRef(ownerReferences []metav1.OwnerReference, ref metav1.OwnerRef
 }
 
+// ReconcileOwnerRef re-parents an object from one OnwerReference to another
+func ReconcileOwnerRef(ownerReferences []metav1.OwnerReference, target metav1.OwnerReference, source metav1.Object) []metav1.OwnerReference {
```
```suggestion
func ReplaceOwnerRef(ownerReferences []metav1.OwnerReference, target metav1.OwnerReference, source metav1.Object) []metav1.OwnerReference {
```

vincepri commented on util/util.go:
```diff
@@ -268,6 +268,22 @@ func EnsureOwnerRef(ownerReferences []metav1.OwnerReference, ref metav1.OwnerRef
+// ReconcileOwnerRef re-parents an object from one OnwerReference to another
+func ReconcileOwnerRef(ownerReferences []metav1.OwnerReference, target metav1.OwnerReference, source metav1.Object) []metav1.OwnerReference {
+	fi := -1
+	for index, r := range ownerReferences {
```
Couldn't we use `indexOwnerRef` here and replace at index?

sethp-nr replied:
`indexOwnerRef` has different semantics, I wanted UID matching to avoid stealing ownership of an object that *should* be deleted.

randomvariable replied:
@sethp-nr Reached any conclusions on this one?

sethp-nr replied:
Oh, sorry, I missed Vince's comment – uniqueness is not guaranteed on object namespace + name across time. If a user deletes a KCP without orphaning dependents and then re-creates it, the intent I think we should honor is to delete the old machines and create a new control plane.

detiber commented on util/util.go:
```diff
@@ -294,9 +310,9 @@ func referSameObject(a, b metav1.OwnerReference) bool {
 
 // PointsTo returns true if any of the owner references point to the given target
-func PointsTo(refs []metav1.OwnerReference, target *metav1.ObjectMeta) bool {
+func PointsTo(refs []metav1.OwnerReference, target metav1.Object) bool {
```
PointsTo should likely use similar logic to `referSameVersion` above, since UID cannot be relied upon when taking pivoting into account.

detiber commented on controlplane/kubeadm/controllers/kubeadm_control_plane_controller.go:
```diff
@@ -834,3 +845,145 @@ func (r *KubeadmControlPlaneReconciler) ClusterToKubeadmControlPlane(o handler.M
+		newSpec := controlplanev1.KubeadmControlPlaneSpec{
+			Version:                machineKubernetesVersion,
+			InfrastructureTemplate: kcp.Spec.InfrastructureTemplate,
+		}
```
Are there associated docs for the adoption process? Considering that most real world providers would require at least a minimum of a JoinConfig in order for additional upgrades or scale operations to work, this makes me feel that we'd trigger an immediate upgrade post-adoption.

sethp-nr replied:
I wrote some in the commit message about the expectation here, where do you think would be a more effective home for that detail?

There's definitely more work to do here, though. The idea is that we're trying to work backwards from the machine to get a valid KCP spec that could have generated it, defaulting to the existing KCP where that's not possible (because "do nothing" is less surprising than "change everything"). It looks like I missed the transition where KubeadmConfigSpec was taken into consideration when hashing the KCP, so we'll have to look at how that should work.

detiber commented on controlplane/kubeadm/controllers/kubeadm_control_plane_controller.go:
```diff
@@ -69,7 +70,7 @@ const (
 
 // +kubebuilder:rbac:groups=core,resources=events,verbs=get;list;watch;create;patch
-// +kubebuilder:rbac:groups=core,resources=secrets,verbs=get;list;watch;create;patch
+// +kubebuilder:rbac:groups=core,resources=secrets,verbs=get;list;watch;create;update;patch
```
Switching to patch should remove the need for adding update to the rbac here>

detiber commented on controlplane/kubeadm/controllers/kubeadm_control_plane_controller.go:
```diff
@@ -834,3 +845,145 @@ func (r *KubeadmControlPlaneReconciler) ClusterToKubeadmControlPlane(o handler.M
+	// See https://github.com/kubernetes/kubernetes/issues/42639
+	uncached := controlplanev1.KubeadmControlPlane{}
+	err := r.uncachedClient.Get(ctx, client.ObjectKey{Namespace: kcp.Namespace, Name: kcp.Name}, &uncached)
+	if err != nil {
```
nit
```suggestion
	if err := r.uncachedClient.Get(ctx, client.ObjectKey{Namespace: kcp.Namespace, Name: kcp.Name}, &uncached); err != nil {
```

detiber commented on controlplane/kubeadm/controllers/kubeadm_control_plane_controller.go:
```diff
@@ -834,3 +845,145 @@ func (r *KubeadmControlPlaneReconciler) ClusterToKubeadmControlPlane(o handler.M
+		ref := m.Spec.Bootstrap.ConfigRef
+		obj := &bootstrapv1.KubeadmConfig{}
+		err := r.Client.Get(ctx, client.ObjectKey{Name: ref.Name, Namespace: kcp.Namespace}, obj)
+		if err != nil {
```
nit:
```suggestion
		if err := r.Client.Get(ctx, client.ObjectKey{Name: ref.Name, Namespace: kcp.Namespace}, obj); err != nil {
```

detiber commented on controlplane/kubeadm/controllers/kubeadm_control_plane_controller.go:
```diff
@@ -834,3 +845,145 @@ func (r *KubeadmControlPlaneReconciler) ClusterToKubeadmControlPlane(o handler.M
+		}
+
+		err = r.AdoptOwnedSecrets(ctx, kcp, obj)
+		if err != nil {
```
nit:
```suggestion
		if err := r.AdoptOwnedSecrets(ctx, kcp, obj); err != nil {
```

detiber commented on controlplane/kubeadm/controllers/kubeadm_control_plane_controller.go:
```diff
@@ -834,3 +845,145 @@ func (r *KubeadmControlPlaneReconciler) ClusterToKubeadmControlPlane(o handler.M
+	secrets := corev1.SecretList{}
+	err := r.Client.List(ctx, &secrets, client.InNamespace(kcp.Namespace))
+
+	if err != nil {
```
nit:
```suggestion
	if err := r.Client.List(ctx, &secrets, client.InNamespace(kcp.Namespace)); err != nil {
```

detiber commented on controlplane/kubeadm/controllers/kubeadm_control_plane_controller.go:
```diff
@@ -834,3 +845,145 @@ func (r *KubeadmControlPlaneReconciler) ClusterToKubeadmControlPlane(o handler.M
+		}, currentOwner))
+
+		err := r.Client.Update(ctx, &ss)
+		if err != nil {
```
nit:
```suggestion
		if err := r.Client.Update(ctx, &ss); err != nil {
```

detiber commented on controlplane/kubeadm/internal/machine_filters.go:
```diff
@@ -93,10 +94,34 @@ func OwnedControlPlaneMachines(controlPlaneName string) MachineFilter {
+// Usage: managementCluster.GetMachinesForCluster(ctx, cluster, AdoptableControlPlaneMachines(cluster.Name, controlPlane))
+func AdoptableControlPlaneMachines(clusterName string) func(machine *clusterv1.Machine) bool {
+	return And(
+		func(machine *clusterv1.Machine) bool {
```
I'm wondering if it would make sense to add new filters to ease the use of this:
```go
func IsNil(machine *clusterv1.Machine) bool {
	return machine == nil
}

func HasControllerRef(machine *clusterv1.Machine) bool {
	return metav1.GetControllerOf(machine) != nil
}
```

Then the test here could be:
```go
return And(
	Not(IsNil),
	ControlPlaneMachines(clusterName),
	Not(HasControllerRef),
)
```

sethp-nr replied:
I can do this. At some point we're fighting the language a bit and causing a fair amount of allocation / pointer hopping, but since we're going to be filtering collections of maybe hundreds of things it's not going to matter :)

sethp-nr commented on controlplane/kubeadm/internal/machine_filters.go:
```diff
@@ -93,10 +94,34 @@ func OwnedControlPlaneMachines(controlPlaneName string) MachineFilter {
 			return false
 		}
-		return controllerRef.Kind == "KubeadmControlPlane" && controllerRef.Name == controlPlaneName
+		return controllerRef.Kind == "KubeadmControlPlane" && controllerRef.Name == owner.GetName() && controllerRef.UID == owner.GetUID()
```
That is a very good point. I don't think the adoption code _will_ reconcile the UID properly as it's currently written, but maybe `clusterctl move` knows how to handle that already? Otherwise, I'd expect the GC to cause trouble for pivoting anyway.

sethp-nr replied:
It appears that clusterctl move does update owner references as objects are moved across: https://github.com/kubernetes-sigs/cluster-api/blob/5154da4caf18b291b1e1e1aba0c635e1d76d3093/cmd/clusterctl/client/cluster/mover.go#L458

That is an impressive piece of code!

I think what that means for us is that UIDs are valid in post-clusterctl v2 workflows. Is there a support matrix where we might need to consider a clusterctl v1 pivot with a KCP?

detiber replied:
I don't think we necessarily need to consider the v1 pivot, but I do think that we need to consider the use case of someone backing up and restoring resources from backup, which I'm not sure that tools in that space would necessarily do the right thing with the UID there.

It seems like the controller-runtime code handles overwriting the controller ref if the api version or the UID are out of date, and I suspect we should likely do the same (even if part of a different workflow than the comparison here)

vincepri commented:
/milestone v0.3.x

Bumping from v0.3.0 as discussed in slack https://kubernetes.slack.com/archives/C8TSNPY4T/p1583780905252500

sethp-nr commented:
With full understanding that it's release day and this got bumped, I'd appreciate feedback on 1ede494 as a way to handle the `KubeadmConfigSpec` field.

I'm still investigating why the e2e test didn't fail, but I think this PR is functionally complete so I'm going to remove the hold.

/hold cancel

sethp-nr commented:
Nope, it's not handling the join config properly.

/hold

vincepri commented:
/milestone v0.3.1

sethp-nr commented:
Latest bits are up, and it does the right thing if we remove the KCP defaulting webhook. Elsewise, I'm seeing some unfortunate (&, it looks, unintentional) changes to the KCP's spec:
```
    kubeadmConfigSpec:
      clusterConfiguration:
        apiServer: {}
        certificatesDir: ""
        controlPlaneEndpoint: ""
        controllerManager: {}
        dns: {}
        etcd: {}
        imageRepository: ""
        kubernetesVersion: ""
        networking:
          podSubnet: ""
        scheduler: {}
      initConfiguration:
        localAPIEndpoint:
          advertiseAddress: ""
          bindPort: 0
        nodeRegistration:
          kubeletExtraArgs:
            eviction-hard: nodefs.available<0%,nodefs.inodesFree<0%,imagefs.available<0%
```

(this from the KCP in test/infrastructure/docker/examples/simple-cluster.yaml)

sethp-nr commented:
/test pull-cluster-api-capd-e2e

chuckha commented on controlplane/kubeadm/controllers/kubeadm_control_plane_controller.go:
```diff
@@ -855,3 +867,131 @@ func (r *KubeadmControlPlaneReconciler) ClusterToKubeadmControlPlane(o handler.M
+			return err
+		}
+
+		if err = controllerutil.SetControllerReference(kcp, m, r.scheme); err != nil {
```
the overall codebases tends to use this style, `if err := ...; err != nil {` when the fn returns only an error. might be good to clean this stuff up but definitely not blocking

chuckha commented on controlplane/kubeadm/controllers/kubeadm_control_plane_controller.go:
```diff
@@ -855,3 +867,131 @@ func (r *KubeadmControlPlaneReconciler) ClusterToKubeadmControlPlane(o handler.M
+			BlockOwnerDeletion: pointer.BoolPtr(true),
+		}, currentOwner))
+
+		if err := r.Client.Update(ctx, ss); err != nil {
```
this should likely be a patch instead of an update, but i think @detiber already mentioned that

chuckha replied:
addressed in a comment that Patch uses a merge strategy that this type doesn't want

sethp-nr commented:
I've:

1. Added docs (thanks Chuck for getting that rolling!)
2. Removed the UID check from OwnedMachines. I didn't change the `PointsTo` semantics, because that's dealing with changing ownership on the cluster's PKI secrets. Nothing was reconciling those owner references previously, and I didn't want to introduce a new behavior where they accidentally "come back to life"
3. Convinced myself that the defaulting behavior cited above _looks_ weird, but doesn't affect the hashing so isn't _wrong_. I plan to continue working on that as part of https://github.com/kubernetes-sigs/cluster-api/issues/1941
4. Rebased

I am interested in pursuing (and/or reviewing) the refactoring that Chuck mentioned, but given the size of this PR I wonder if it wouldn't be better to do that work in a follow up?

/hold cancel

sethp-nr commented:
Thanks for the review! Good call on providing examples of what's not going to work – I reworded those bits to hopefully make more sense.

re Patch vs Update: The object being Updated here is the `Secret` that contains the PKI materials, and the trouble is that it's already owned. I don't believe strategic merge offers "replace" or "delete" semantics, so in order to modify the existing ownership I'd have to resort to a RFC6902 patch, which the patch helper currently doesn't know anything about.

In this case, there didn't seem to be too much danger because it seems no one else is writing these secrets after cluster init (and it being a core type make it less likely we'll fight over schema changes with an out-of-date controller anyway). If we _did_ get a read-modify-write race (AKA "resourceVersion out of date") we'd re-start the adoption process because we're not setting the owner ref on the Machine until after the secrets are successfully adopted.

chuckha commented:
I'm happy with this, thanks for all the updates!

/approve
/assign @vincepri 

assigning to vince for lgtm

vincepri commented:
Currently reviewing

vincepri commented on bootstrap/kubeadm/api/equality/semantic.go:
```diff
@@ -0,0 +1,127 @@
+limitations under the License.
+*/
+
+package equality
```
Are these only used for adoption path? I'm wondering if we should put them under the internal package and drop them in v1alpha4

sethp-nr replied:
Right now, yes.

The question is: what are we going to do for v1a4 to avoid dropping a new KCP controller next to an existing v1a3 control plane and having it go on and "upgrade" every CP machine it sees until the hashes match?

The way the deployment controller solves this problem upstream is by eschewing the hash for comparison and using a "semantic equals" on the pod templates. That's the route I imagined taking with this package.

vincepri replied:
I'm not 100% if i'm following, but I was hoping we could drop adoption when v1alpha4 is released and have one true way to spin up a control plane for cluster api

sethp-nr replied:
Ah, this is both at 90 degrees to the change and a longer conversation. Even w/o adoption we'd still want to consider the semantic equals thing to make it possible to change the KCP type in the future, and I have some examples about why adoption is useful outside of migration contexts.

vincepri commented:
/milestone v0.3.x

resetting the milestone pending comments above

sethp-nr commented on controlplane/kubeadm/controllers/kubeadm_control_plane_controller.go:
```diff
@@ -858,3 +870,131 @@ func (r *KubeadmControlPlaneReconciler) ClusterToKubeadmControlPlane(o handler.M
+			continue
+		}
+
+		ss := s.DeepCopy()
```
We're about to modify it, so I think so – would you rather we replace or shadow the loop variable instead?

vincepri replied:
either way should be ok, I might have missed the loop while reviewing

sethp-nr commented on util/util.go:
```diff
@@ -426,3 +443,16 @@ func (o MachinesByCreationTimestamp) Less(i, j int) bool {
 	return o[i].CreationTimestamp.Before(&o[j].CreationTimestamp)
 }
+
+// IsSupportedVersionSkew will return true if a and b are no more than one minor version off from each other
```
I don't think so – this is an in-cluster skew: https://kubernetes.io/docs/setup/release/version-skew-policy/#supported-version-skew

> newest kube-apiserver is at 1.13
other kube-apiserver instances are supported at 1.13 and 1.12

vincepri replied:
Ah ok makes sense

sethp-nr commented on controlplane/kubeadm/internal/machinefilters/machine_filters.go:
```diff
@@ -56,6 +58,16 @@ func Not(mf Func) Func {
+// HasControllerRef is a filter that returns true if the machine has a controller ref
+var HasControllerRef = func(machine *clusterv1.Machine) bool {
+	return metav1.GetControllerOf(machine) != nil
+}
```
😆because of copy pasta, and me forgetting how go works

sethp-nr commented on controlplane/kubeadm/internal/machinefilters/machine_filters.go:
```diff
@@ -93,10 +105,32 @@ func OwnedControlPlaneMachines(controlPlaneName string) Func {
+		Not(IsNil),
+		ControlPlaneMachines(clusterName),
+		Not(HasControllerRef),
+	)
```
That's being checked before adoption in the controller, along with the version.

I figured it would be better in those cases to have the KCP controller emit an "I don't know what to do" error in a loop than proceed as if the existing CP machines weren't there.

vincepri replied:
I was thinking this could impact the reconciliation of a KCP instance if there is some user mistake. Let's say for example, some user (or controller) creates a machine with a control-plane label, KCP controller will be stuck in trying to adopt the machines

sethp-nr replied:
Yeah, that's what I think I would expect as a user: if I make up a control plane node, I would expect the KCP to either take it over or tell me why it can't.

The alternative I saw a few times in testing was that it would happily spin up a second control plane, with the same certs, behind the same LB, but backed by a different etcd.

sethp-nr commented on controlplane/kubeadm/internal/machinefilters/machine_filters.go:
```diff
@@ -93,10 +105,32 @@ func OwnedControlPlaneMachines(controlPlaneName string) Func {
+// Usage: managementCluster.GetMachinesForCluster(ctx, cluster, AdoptableControlPlaneMachines(cluster.Name, controlPlane))
+func AdoptableControlPlaneMachines(clusterName string) func(machine *clusterv1.Machine) bool {
+	return And(
+		Not(IsNil),
```
`HasControllerRef` doesn't check.

vincepri replied:
Should we check in there instead of having the `IsNil` filter? I'd try to keep it consistent

sethp-nr commented on controlplane/kubeadm/internal/machinefilters/machine_filters.go:
```diff
@@ -93,10 +105,32 @@ func OwnedControlPlaneMachines(controlPlaneName string) Func {
 			return false
 		}
-		return controllerRef.Kind == "KubeadmControlPlane" && controllerRef.Name == controlPlaneName
+		return controllerRef.Kind == "KubeadmControlPlane" && controllerRef.Name == owner.GetName()
```
That's comparing two refs, whereas here we have a ref and an object. I'll steal the important bits and rework it.

sethp-nr commented on controlplane/kubeadm/controllers/kubeadm_control_plane_controller.go:
```diff
@@ -858,3 +870,131 @@ func (r *KubeadmControlPlaneReconciler) ClusterToKubeadmControlPlane(o handler.M
+		// 0. get machine.Spec.Version - the easy answer
+		machineKubernetesVersion := ""
+		if m.Spec.Version != nil {
+			machineKubernetesVersion = *m.Spec.Version
```
We'll have already blown up trying to parse the version above during the safety check. Rather than trying to chase down what exactly the version of code is running (a hard problem!) this is just a "good enough" check that mostly does the right thing.

sethp-nr commented on controlplane/kubeadm/controllers/kubeadm_control_plane_controller.go:
```diff
@@ -858,3 +870,131 @@ func (r *KubeadmControlPlaneReconciler) ClusterToKubeadmControlPlane(o handler.M
+		}
+
+		if !util.IsSupportedVersionSkew(kcpVersion, machineVersion) {
+			r.recorder.Eventf(kcp, corev1.EventTypeWarning, "AdoptionFailed", "Could not adopt Machine %s/%s: its version (%q) is outside supported +/- one minor version skew from KCP's (%q)", m.Namespace, m.Name, *m.Spec.Version, kcp.Spec.Version)
```
I think the recorder will log "for real" – are you thinking for tests?

vincepri commented:
@sethp-nr @rudoi thoughts on getting this rebased / cleaned up for v0.3.4?

sethp-nr commented:
I'm working on the rebase as I catch back up this week. It's a tad rough, ngl :)

sethp-nr commented:
Ok @vincepri this is RFR again – maybe even for the last time :)

vincepri commented:
@sethp-nr now that we have the new clusterctl test framework in place, can we move the e2e tests to that one instead?

randomvariable commented:
Finished with manual testing. Happy with the output.

We need to visit the breaking API changes though.

vincepri commented:
The `pull-cluster-api-capd-e2e ` is failing because the job has been removed, @sethp-nr is this PR rebased on master?

vincepri commented:
```
sigs.k8s.io/cluster-api/test/e2e
  Compatible changes:
  - HaveControllerRef: added
  - KCPAdoptionSpec: added
  - KCPAdoptionSpecInput: added
sigs.k8s.io/cluster-api/test/framework/exec
  Incompatible changes:
  - KubectlApply: changed from func(context.Context, string, []byte) error to func(context.Context, string, []byte, ...string) error
sigs.k8s.io/cluster-api/util/secret
  Compatible changes:
  - DefaultCertificatesDir: added
sigs.k8s.io/cluster-api/api/v1alpha3
  Compatible changes:
  - (*NetworkRanges).String: added
sigs.k8s.io/cluster-api/bootstrap/kubeadm/api/equality
  Compatible changes:
  - SemanticMerge: added
sigs.k8s.io/cluster-api/util
  Incompatible changes:
  - PointsTo: changed from func([]k8s.io/apimachinery/pkg/apis/meta/v1.OwnerReference, *k8s.io/apimachinery/pkg/apis/meta/v1.ObjectMeta) bool to func(k8s.io/apimachinery/pkg/apis/meta/v1.Object, Object) bool
  Compatible changes:
  - IsControlledBy: added
  - IsSupportedVersionSkew: added
  - Object: added
  - ReplaceOwnerRef: added
sigs.k8s.io/cluster-api/test/framework
  Incompatible changes:
  - ClusterProxy.Apply: changed from func(context.Context, []byte) error to func(context.Context, []byte, ...string) error
  - ClusterProxy: no longer implements Applier
  - TypeToKind: changed from func(interface{}) string to func(k8s.io/apimachinery/pkg/runtime.Object) string
  Compatible changes:
  - WaitForClusterMachineNodeRefs: added
  - WaitForClusterMachineNodeRefsInput: added
  - WaitForClusterMachinesReady: added
  - WaitForClusterMachinesReadyInput: added
  - WaitForControlPlaneToBeUpToDate: added
  - WaitForControlPlaneToBeUpToDateInput: added
+ EXIT_VALUE=1
```

vincepri commented:
I'd say we should probably avoid the incompatible changes, definitely let's squash and rebase on master to make sure we're not getting changes that already happened upstream

sethp-nr commented:
I can rebase past the e2e removal easily enough. 

For api changes, it sounds like I should be adding new methods & deprecating old ones?

vincepri commented:
Yeah that seems the way to go

sethp-nr commented:
Ok, I introduced new methods in `891f4e8cb` – if that looks ok, I'll squash it in.

randomvariable commented:
Looking good.

vincepri commented:
Doing another pass

vincepri commented on util/util.go:
```diff
@@ -388,16 +404,52 @@ func referSameObject(a, b metav1.OwnerReference) bool {
+}
+
+// Returns true if ref refers to obj
+func refersTo(ref *metav1.OwnerReference, obj Object) bool {
```
This internal method and the other one are almost the same :/

sethp-nr replied:
Yeah, interface vs struct in the second position. I'll move 'em next to each other at least.

vincepri commented on util/util.go:
```diff
@@ -388,16 +404,52 @@ func referSameObject(a, b metav1.OwnerReference) bool {
 	return false
 }
 
+type Object interface {
+	metav1.Object
+	GroupVersionKind() schema.GroupVersionKind
+}
```
Can we make this interface private? There is https://github.com/kubernetes-sigs/controller-runtime/pull/898 upstream that we might be able to use at some point, although I don't want to expose this interface just yet (and make it an API)

vincepri commented on util/util.go:
```diff
@@ -388,16 +404,52 @@ func referSameObject(a, b metav1.OwnerReference) bool {
+	return false
+}
 
+// PointsToObject returns true if any of the owner references point to the given target
+func PointsToObject(obj metav1.Object, target Object) bool {
```
`PointsToObject` is a little too generic, should this be `IsOwnedByObject`?

vincepri commented on util/util.go:
```diff
@@ -546,3 +598,16 @@ func ClusterToObjectsMapper(c client.Client, ro runtime.Object, scheme *runtime.
+	if a.Major != b.Major {
+		return false
+	}
+
```
```suggestion
```

sethp-nr commented on controllers/cluster_controller_test.go:
```diff
@@ -703,6 +703,10 @@ func TestFilterOwnedDescendants(t *testing.T) {
 	g := NewWithT(t)
 
 	c := clusterv1.Cluster{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: clusterv1.GroupVersion.String(),
+			Kind:       "Cluster",
+		},
```
Because `PointsToObject` checks Kind and group (though not version) per Jason's suggestion to have it match on the object kind / group / name triple.

vincepri commented on util/util.go:
```diff
@@ -372,6 +391,42 @@ func indexOwnerRef(ownerReferences []metav1.OwnerReference, ref metav1.OwnerRefe
+	return false
+}
+
+type object interface {
+	metav1.Object
+	GroupVersionKind() schema.GroupVersionKind
+}
```
If we rebase on master, we should be able to consume `controllerutil.Object` from https://github.com/kubernetes-sigs/controller-runtime/pull/898

vincepri commented:
/milestone v0.3.7

vincepri commented:
/lgtm
